src/Iterators.jl

```diff
@@ -894,4 +895,94 @@ macro chain(ex)
     Expr(:let, Expr(:block, cycleex...), states...)
 end
 
+
+type MergedIter
+    it1   # iterator 1
+    it2   # iterator 2
+    lt    # custom `<` operator to compare elements of 2 iterators    
+end
+
+MergeIter(it1, it2) = MergeIter(it1, it2, isless)
+
+type MergedState
+    hd1   # current head of 1st iterator
+    hd2   # current head of 2nd iterator
+    s1    # current state of 1st iterator
+    s2    # current state of 2nd iterator
+end
+
+
+function start(it::MergedIter)
+    s1 = start(it.it1)
+    s2 = start(it.it2)
+    # doing 1 step ahead and caching next element of each iter
+    hd1, s1 = safenext(it.it1, s1)
+    hd2, s2 = safenext(it.it2, s2)
+    return MergedState(hd1, hd2, s1, s2)
+end
+
+"""
+Like `next`, but may be called on iterator that is already done. In this case
+new item is `nothing` and new state is the same as previous. 
+"""
+function safenext(it, s)
+    if done(it, s)
```

Comment:
I had it working like this at the beginning and the code looked much more complicated. The trick here is that at each step of merged iterator we actually hold 2 steps of each underlying iterator. Using `complete` and `safenext` makes it possible to treat them almost like normal iterators. E.g. you cannot call `next` on a file iterator that is already `done`, but you can call `safenext()` instead. This also plays well with `nothing` for a head element which indicates that underlying iterator is not only `done`, but also doesn't have "cached" head element. 

I have an idea how to handle it using a kind of `PrefetchIteratorWrapper` - wrapper that behaves like an iterator, but internally prefetches 1 element in advance and hides all the dirty details. I'll give it a try, but honestly I'm not sure if it will make code better or worse.